Scoring and typing

ZedXC opened the pull request:
a new minigame + attention bar and chances to how the game scores and functions around ending games and beginning new ones.

ZedXC committed:
Typing Minigame and Bug fixes

-Added base for typing minigame
-Fixed bugs with typing minigame

ZedXC committed:
Attention Bar added. Doesn't do anything atm

ZedXC committed:
Merge branch 'main' into scoring-and-typing

ZedXC committed:
scoring with multiplies and attention meter

Implemented:
- Scoring
- Attention bar that can be increased and decreased
- New way to pass information about how well game has gone
- changes to juggling and typing games to reflect new scoring

ZedXC committed:
small name change

ZedXC committed:
Merge branch 'main' into scoring-and-typing

tessapower commented on jest-to-impress/scenes/mini_games/typ5A4C.tmp:
This temp file should be deleted.

tessapower commented on jest-to-impress/scenes/mini_games/typ6DF3.tmp:
This temp file should be deleted.

tessapower commented on jest-to-impress/scripts/game.gd:
```diff
@@ -32,36 +43,52 @@ func _unhandled_input(event) -> void:
 
 # A callback function intended to be called by a mini-game when the player loses
 func on_failure() -> void:
+  decrease_attention_meter(10)
```
Godot is complaining about incorrect indentation here, so it stops the game from running.

ZedXC replied:
ah that was when I fixed the merge conflicts. easy solve

tessapower commented on jest-to-impress/scripts/game.gd:
```diff
@@ -80,3 +107,23 @@ func next_mini_game() -> void:
 	# Load the mini-game at the new index
 	load_mini_game(current_idx)
+	
+func increase_score(increase : int) -> void:
+	GamestateManager.increase_score(1)
+	if attention_meter_instance.get_attention() + increase > attention_meter_max:
+		set_attention_meter(attention_meter_max)
+		GamestateManager.score_multipier += 1
+	else:
+		increase_attention_meter(increase)
+		GamestateManager.score_multipier = 1
+	
+func increase_attention_meter(attention_change : int) -> void:
+	var attention_meter_change = attention_meter_instance.get_attention() + attention_change
+	attention_meter_instance.set_attention(attention_meter_change)
+
+func decrease_attention_meter(attention_change : int) -> void:
+	var attention_meter_change = attention_meter_instance.get_attention() - attention_change
+	attention_meter_instance.set_attention(attention_meter_change)
+
+func set_attention_meter(attention_change : int) -> void:
+	attention_meter_instance.set_attention(attention_change)
```
These functions should really be removed in favor of the attention meter data being stored in the Gamestate Manager and the attention meter scene itself requesting that data from the Gamestate Manager to know what to display at any given point in time. Doing so will keep all player related data in one place, and the logic for when the game is over in the logical place.

This will also make it easier for each mini-game to update the score and attention meter appropriately by going through the Gamestate Manager, also removing the need for the `on_finished` function in the `game` script to do anything other than load the next random mini-game.

Seeing as we only want update the attention meter based on the player's performance, we would ideally have a function that takes an enum and uses a match statement (Godot's version of a switch) to do this, calling this function at the appropriate time in the mini-game's lifecycle from the mini-game's own script - this will allow each mini-game to decide the right way to evaluate performance.

tessapower commented on jest-to-impress/scripts/mini_games/typingMinigame.gd:
We need to change how the player pauses the game, possibly by adding a button, as typing 'p' pauses the game so I can't type any punchline that contains a 'p'.

tessapower replied:
I'm unsure how it happened, but I can only play this mini-game once on the first play through, and every subsequent play through this mini-game doesn't show up.

ZedXC replied:
Could we change the pause button to be esc instead of P? I feel like thats more intuitive?

tessapower reviewed (changes requested):
Some architectural changes and a few bugs to look into!

ZedXC committed:
Improvements to the attention meter

- Made attention meter information be stored in the gameStateManager.
- Deleted temp files
- Fixed indentation issues
- changed the on_finished function to only load the next random minigame
- Added function to allow for changes to the attention meter while game is still running.
- Changed the pause button from p to esc

ZedXC committed:
Allowed for input of ENUMS

ZedXC committed:
Merge branch 'main' into scoring-and-typing

ZedXC committed:
Merge branch 'main' into scoring-and-typing

tessapower committed:
Merge branch 'main' into scoring-and-typing

tessapower committed:
Fix bug with high scores

tessapower committed:
Rename files to conform to used naming convention

tessapower committed:
Refactor Gamestate Manager to reduce attention meter code

tessapower committed:
Fixup: Rename files to conform to used naming convention

tessapower committed:
Refactor mini-games to update attention meter only

tessapower committed:
Add better comments to knife-throwing

tessapower committed:
Remove file artefact

tessapower committed:
Fix bugs in typing mini-game

tessapower committed:
Fix whatever I broke in the juggling game
